Pricing grid: Implement tooltips on mobile

niranjan-uma-shankar opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 2002-gh-martech

## Proposed Changes

This PR fixes a bug that tooltips in the pricing grid were not displayed in mobile viewports. 

<img width="297" alt="Screenshot 2023-08-03 at 11 49 16 AM" src="https://github.com/Automattic/wp-calypso/assets/1269602/1f91cb4c-b572-4112-97d1-a589dd16c985">

#### Note

When I initially tried managing the state for the tooltips in the parent component `<PlanFeatures2023Grid />` in [this file](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plan-features-2023-grid/index.tsx),  an issue I faced is that whenever I clicked on a feature, the Foldable Card will close. This was because clicking on a feature will set a state variable to make the tooltip visible, which in turn re-renders the `<PlanFeatures2023Grid />` and its child `<FoldableCard/>` resets its `expanded` value to its initial value of false. So, I had to move the plan features and plan logos into a separate component so that the re-renders are contained within the child component. 




## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Signup Flow

1. Go to `/start`:
    * Verify in desktop viewport that you can see tooltips in both the main pricing grid and below-the-fold comparison table, when you mouse over a fearture. The tooltip should hide when the mouse pointer oves away. 
    * Verify in mobile viewport that you can see a tooltip when you click a feature on a plan card, Jetpack logo in the feature list, plan logo on the Business / Commerce / Enterprise plan. Clicking on the same item again should hide the tooltip. Clicking anywhere on a plan card when a tooltip is already open should hide it. 

![tooltip](https://github.com/Automattic/wp-calypso/assets/1269602/0ea48894-9d24-4f02-bd0b-384fdd604d82)

    
2. In a logged-in account, visit Upgrades > Plans and test everything mentioned above for `/start`. If you are on a paid plan, then clicking on the "Downgrade" button will show a tooltip in Desktop but not in Mobile (since it already includes text about the Downgrade, so a separate tooltip is not necessary). 


3. Check other signup flows like newsletter signup flow and verify that the plans step shows tooltips in mobile.

niranjan-uma-shankar committed:
Implement tooltips

niranjan-uma-shankar committed:
Close all tooltips if clicked outside

niranjan-uma-shankar commented on client/my-sites/plan-features-2023-grid/components/plan-logo.tsx:
```diff
@@ -0,0 +1,99 @@
+						) }
+						id="wp-cloud-logo"
+						setActiveTooltipId={ setActiveTooltipId }
+						activeTooltipId={ activeTooltipId }
```
This new file simply holds the PlanLogo component which existed in client/my-sites/plan-features-2023-grid/index.tsx. The only additions here are the additional args `setActiveTooltipId` and `activeTooltipId` for the `<Plans2023Tooltip />` component.

niranjan-uma-shankar commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -139,77 +129,6 @@ type PlanFeatures2023GridState = {
 	showPlansComparisonGrid: boolean;
 };
 
-const PlanLogo: React.FunctionComponent< {
```
As mentioned above, this has been moved to its own file for handling component state handling.

niranjan-uma-shankar commented on client/my-sites/plan-features-2023-grid/components/plan-features.tsx:
```diff
@@ -0,0 +1,103 @@
+	const [ activeTooltipId, setActiveTooltipId ] = useState( '' );
+	const isDesktop = useDesktopBreakpoint();
+
+	useEffect( () => {
```
This file holds the features as a separate child component which previously existed in client/my-sites/plan-features-2023-grid/index.tsx. The addition here is the tooltip state handler and useEffect to determine the currently active tooltip.

niranjan-uma-shankar commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -848,47 +784,19 @@ export class PlanFeatures2023Grid extends Component<
 				! isWooExpressPlusPlan( properties.planName )
 		);
 
-		return planProperties
```
This block is moved to client/my-sites/plan-features-2023-grid/components/plan-features.tsx above.

niranjan-uma-shankar committed:
Fix type

niranjan-uma-shankar committed:
Some code comments

niranjan-uma-shankar committed:
Tooltip handler for downgrade

niranjan-uma-shankar committed:
Remove downgrade tooltip on mobile

aneeshd16 commented on client/my-sites/plan-features-2023-grid/components/plan-features.tsx:
```diff
@@ -0,0 +1,105 @@
+	Container,
+} ) => {
+	const [ activeTooltipId, setActiveTooltipId ] = useState( '' );
+	const isDesktop = useDesktopBreakpoint();
```
The plans grid uses different breakpoints from the rest of Calypso, so I think using `useDesktopBreakpoint` would cause issues at certain breakpoints:
https://github.com/Automattic/wp-calypso/blob/178ed82ccf2a0b094e6d60d0526184d378126a36/client/my-sites/plan-features-2023-grid/_media-queries.scss#L75-L80
https://github.com/Automattic/wp-calypso/blob/8d3ad9ee07de9d7e55ae2be42c5b3e26069e535d/client/my-sites/plan-features-2023-grid/_media-queries.scss#L6-L8

We're figuring out ways to improve this, here's the WIP PR: https://github.com/Automattic/wp-calypso/pull/79129.

I don't have any good solutions to move this PR forward, maybe something like `isDesktop = useBreakpoint('>780px')`?

niranjan-uma-shankar replied:
I realised that I don't need the breakpoints at all. I was using it to determine whether or not to add the touch event listeners, and I made a change to check for whether the browser supports touch events instead of relying on viewport size.

chriskmnds reviewed (commented):
@aneeshd16 left a comment that sounds like a follow up is expected. From my end, I'd appreciate if this PR gets rebased onto #79733 since the refactors included here will lead to several conflicts there. #79733 has proven especially difficult for us

niranjan-uma-shankar committed:
Fix merge conflicts

niranjan-uma-shankar committed:
Add touch listener if browser support touch events only

niranjan-uma-shankar committed:
Clean up

niranjan-uma-shankar committed:
Expand the JP logo clickable area

niranjan-uma-shankar committed:
Expand the JP logo clickable area

niranjan-uma-shankar committed:
Expand the JP logo clickable area

niranjan-uma-shankar committed:
Revert changes to expand JP logo area

niranjan-uma-shankar commented:
> Not at all a blocker for this, but it's a little strange how the tooltip positions tend to vary so much, from a vertical alignment perspective. Something to look into at another time.
Maybe it's just in the browser emulator though – I haven't tested on a real mobile device.

It's weird indeed, and I'd need to debug the tooltip component to understand why the positioning is off. I agree that it's not a blocker but I'm tracking this issue in 2056-gh-martech.

> The touchpoint on the Jetpack logo seems a little off. Again, maybe it's just the browser emulator. Not a blocker.

The logo is small and the touch area somehow is not enveloping the entire image area. Also tracking this in 2056-gh-martech.

> In the newsletter flow (http://calypso.localhost:3000/setup/newsletter/plans), I noticed a couple features with missing tooltips. There might be others, I didn't test them all. Probably out of scope for this task, but I wanted to mention it.

Thanks, I've seen that in the newsletter flow earlier - it does not have the tooltip content for these features.

niranjan-uma-shankar commented:
@travisw @aneeshd16 @chriskmnds I've rebased the PR, could you please review this again when you get a chance?

niranjan-uma-shankar committed:
fix TS errors

niranjan-uma-shankar committed:
Fix TS error

niranjan-uma-shankar committed:
Fix TS error

southp commented on client/my-sites/plan-features-2023-grid/components/plan-features.tsx:
```diff
@@ -0,0 +1,112 @@
+		// Close all toolltips in mobile if the user clicks anywhere on the plan card.
+		const closeAllTooltips = ( event: TouchEvent ) => {
+			const eventTarget = event.target as Element;
+			if ( ! eventTarget?.classList.contains( 'plans-2023-tooltip__hover-area-container' ) ) {
```
What this does is that when the user clicks anywhere within this hover area container all the tooltips will be dismissed. The hover area container is a `span` wrapped around the feature string only, so clicking  the blank area won't dismiss the tooltips. Could you help me confirm if this is expected? I'm asking because it's not aligned with the comment above :) 

https://github.com/Automattic/wp-calypso/assets/1842898/b5d9e14a-70a0-412e-9a8d-6f68ce94dde6

niranjan-uma-shankar replied:
> so clicking the blank area won't dismiss the tooltips

This section of the code is entered only on touch devices due to the `isTouch()` condition above, so you'd need to test on an actual mobile device. Touching anywhere on the plan card will dismiss all tooltips since the event target class will not contain `plans-2023-tooltip__hover-area-container`. 

> not aligned with the comment above 

I modified the code comment slightly to rename "click" to "touch" to avoid ambiguities.

Let me know if this makes sense!

southp commented on client/my-sites/plan-features-2023-grid/components/plan-comparison-grid.tsx:
```diff
@@ -667,12 +667,14 @@ const PlanComparisonGridFeatureGroupRow: React.FunctionComponent< {
 	isHighlighted,
 	intervalType,
 } ) => {
+	const [ activeTooltipId, setActiveTooltipId ] = useState( '' );
```
From the code here, it looks like unexpected that touching a feature in the comparison grid doesn't show a tooltip. The production is actually also buggy: in the tablet resolution, hovering a feature won't reveal a tooltip.

niranjan-uma-shankar replied:
I've fixed this. Tooltips now work on the comparison grid in tablets and mobile. Note that the touch behaviour will have to be tested on an actual device.

southp commented on client/my-sites/plan-features-2023-grid/components/actions.tsx:
```diff
@@ -201,6 +202,7 @@ const LoggedInPlansFeatureActionButton = ( {
 	selectedSiteSlug: string | null;
 	planActionOverrides?: PlanActionOverrides;
 } ) => {
+	const [ activeTooltipId, setActiveTooltipId ] = useState( '' );
```
This state hook is duplicated both in here and in the plans comparison grid. Could we extract it into a self-contained hook so the code is [DRY-er](https://en.wikipedia.org/wiki/Don't_repeat_yourself)?

niranjan-uma-shankar replied:
Great point. I've added this fix and moved this code to a custom hook.

niranjan-uma-shankar committed:
Update comment

southp commented on client/my-sites/plan-features-2023-grid/components/plan-features.tsx:
```diff
@@ -0,0 +1,112 @@
+import { JetpackLogo } from '@automattic/components';
+import { LocalizeProps } from 'i18n-calypso';
+import { useState, useEffect } from 'react';
+import { hasTouch } from 'calypso/lib/touch-detect';
```
Since Decepticons are working on extracting the new pricing grid into an isolated NPM package, this will introduce a dependency that we will need to cleanup very soon: You can imagine everything under `plans-features-2023-grid` directory will be moved into `packages`, and calypso will be more like the user side of this package, wiring all the props and configuration inside `PlansFeaturesMain`. Thus, under that structure, `plan-features-2023-grid` won't have access to this library. Could we look into ways of decoupling this dependency? It's just a simple function so an easy way would be just implement the same functionality inside of the libarary of `plan-features-2023-grid` and comment about the reason there. What do you think?

Another option is to extract `touch-detect` into its own NPM package and publish it to npmjs. However, it's copied from Modernizr so I don't feel appropriate to publish it as something from us. Also they [has removed ths functionality](https://github.com/Modernizr/Modernizr/pull/2432) 3 years ago.

Btw, for more details about the NPM extraction project, here is the related tasks: https://github.com/orgs/Automattic/projects/552/views/1?filterQuery=NPM and the reasoning and principles explained by @chriskmnds : pdvytD-bG-p2

niranjan-uma-shankar replied:
I understand this feedback is for an upcoming refactor, but I think moving the touch-detect out of its library is out of scope for this PR. IMHO I feel the package extraction effort to factor in the best way to use this lib (either repeating code within the package, or adding a dependency for an external package). Let me know if you feel strongly about your opinion though, more than happy to revist my comment.

niranjan-uma-shankar committed:
Refactor to DRY code

niranjan-uma-shankar committed:
Update import

niranjan-uma-shankar committed:
Update to a better name for a component

niranjan-uma-shankar committed:
Refactor

niranjan-uma-shankar committed:
Fix bug with plan logo tooltip

niranjan-uma-shankar committed:
Simplify - use closest()

niranjan-uma-shankar committed:
Enable tooltips for the mobile comparison grid

niranjan-uma-shankar reviewed (commented):
Thanks for the feedback and review @southp! I've pushed in a few changes which I'm summarising below:

* Extracted the tooltip state management to a [custom hook](https://github.com/Automattic/wp-calypso/pull/80082/files#diff-4e6a9ce25f440de40cd7ec18f17242c61e2667ed4cd091bdf5e4ff379953f26c).
* Fixed the issue with tooltips not working for the below-the-fold comparison grid in mobile (issue also exists in production)
* Renamed `<PlanFeatures />` component which I introduced in this PR to a better named `<PlanFeaturesContainer />`/
* Removed breakpoint based detection in `<Plans2023Tooltip />` component, and used touch based detection.

### Note

Since I'm using touch based detection, to accurately test the mobile and tablet touch detection, you'd need to test on an actual device, since Chrome's inspect tools will only simulate the browser viewport and the code will still identify your desktop to be a non-touch device.

southp commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -491,11 +406,12 @@ export class PlanFeatures2023Grid extends Component< PlanFeatures2023GridType >
-					isTableCell={ options?.isTableCell }
 					isInSignup={ isInSignup }
+					isTableCell={ options?.isTableCell }
+					Container={ Container }
```
I missed this in the last review –– it doesn't feel right that the Container is passed as one of the props of `<PlanLogo>` for two reasons:

1. The most natural way to express this heirarchical relationship in DOM is simply `<Container><Child></Child></Container>`, which is also how the other places using `<Container>`
1. `Container` is a private JSXElement defined in `plan-features-2023-grid`. By moving this into a prop, `<PlansLogo>` has to duplicate the same typing declaration. 

In case of `<PlanLogo>`, it's relatively simple to just move `Container` out, but the features container would be a bit more involving. More details below.

southp commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -721,45 +637,18 @@ export class PlanFeatures2023Grid extends Component< PlanFeatures2023GridType >
+				selectedFeature={ selectedFeature }
+				isCustomDomainAllowedOnFreePlan={ isCustomDomainAllowedOnFreePlan }
+				isTableCell={ options?.isTableCell }
+				Container={ Container }
```
Following the reasoning above, ideally `Container` shouldn't be passed as a prop here. However, it might be more involving to extract it out since the current `<PlanFeauresContainer/>` encapsulates the `map` call. One potential way is to not extract the `map` and `Container` out, but only the part contained within it, i.e. `<PlanFeatures2023GridFeatures>`, `<Plans2023Tooltip>`, `<JetpackLogo>`, etc.

niranjan-uma-shankar replied:
@southp How do you feel about moving `<Container />` to a separate component in the components folder and importing that in the places where it's used?

southp replied:
Good point. I think that also works :) In that case, it would be better to rename it into something more specific to its usage so whoever work on calypso wouldn't bump into it and thought they've found a generic container component. `Container` is a bit too generic.

niranjan-uma-shankar replied:
Agree 😄

southp reviewed (commented):
@niranjan-uma-shankar Thanks for addressing the feedbacks :) 
I've tested it on a real mobile device, desktop, and simulated mobile devices, and they all work as described –– on my end, somehow the simulated devices also work :) 

I've left a few more feedbacks on the structural issues. Sorry that I missed them in my last round of review 🙇🏼 

Regarding:
> I understand this feedback is for an upcoming refactor, but I think moving the touch-detect out of its library is out of scope for this PR. IMHO I feel the package extraction effort to factor in the best way to use this lib (either repeating code within the package, or adding a dependency for an external package). Let me know if you feel strongly about your opinion though, more than happy to revist my comment.

Yes, I agree that it goes beyond the scope of this PR. However, since doing all the foundational restructuring works while many works on top of it is happenning has been generating numerous , I'd really appreciate if you can save a follow-up issue to tackle for us :) To begin with, I think it's already enough to copy `hasTouch()` to somewhere under `my-sites/plan-features-2023-grid/lib`. Does that make sense to you?

niranjan-uma-shankar committed:
Extract <Container /> to a separate component, and rename to PlanDivOrTdContainer

niranjan-uma-shankar committed:
Extract <Container /> to a separate component, and rename to PlanDivOrTdContainer

niranjan-uma-shankar commented:
@southp I've moved the existing `<Container />` to a separate component `<PlanDivOrTdContainer />`. 

> I'd really appreciate if you can save a follow-up issue to tackle for us :)

Sure, added issue 2088-gh-martech to track this. It's also on the Backlog column of the "2023 Pricing Page Optimization" board.

southp reviewed (approved):
I've retested it and it works as described 🚢
